### PR TITLE
[Kunlun]Multi xpu dygraph performance optimization , add distributed.spawn support for multi xpu and some bug-fixes

### PR DESCRIPTION
**Pull request by vslyu**
### PR types

  Performance optimization 
### PR changes

Others
### Describe

Lists of major modifications and bug-fix about Baidu Kunlun XPU:
1.  Multi xpu dygraph training performance optimization.
     add new threads for multi xpu communication in imperative/reducer 
2. Add distributed.spawn support for multi xpu. [PR31032](https://github.com/PaddlePaddle/Paddle/pull/31032)(merged into this PR)
     register FLAGS_selected_xpus  into pybind/global_value_getter_setter, add xpu interface in distributed/spawn and 
     distributed/utils. example at the end.
3. Export bkcl_comm_num interface for python 
     add  bkcl_comm_num interface in pybind/pybind, and remove the limit of ‘num_threads = 1’   in fluid/compiler.
4. Fix extend device id(>10) to multi strings bug in fleet/launch_utils.
    replace extend with append in fleet/launch_utils.
5. Fix some error info and macro definition in collective/c_comm_init_op and collective/gen_bkcl_id_op.

use environment variables:
```export FLAGS_selected_xpus = 0,1,2,3```
specifies these cards to run.

spawn example code:
```
from __future__ import print_function

import paddle
import paddle.nn as nn
import paddle.optimizer as opt
import paddle.distributed as dist

class LinearNet(nn.Layer):
    def __init__(self):
        super(LinearNet, self).__init__()
        self._linear1 = nn.Linear(10, 10)
        self._linear2 = nn.Linear(10, 1)

    def forward(self, x):
        return self._linear2(self._linear1(x))

def train(print_result=False):
    # 1. initialize parallel environment
    dist.init_parallel_env()

    # 2. create data parallel layer & optimizer
    layer = LinearNet()
    dp_layer = paddle.DataParallel(layer)

    loss_fn = nn.MSELoss()
    adam = opt.Adam(
        learning_rate=0.001, parameters=dp_layer.parameters())

    # 3. run layer
    inputs = paddle.randn([10, 10], 'float32')
    outputs = dp_layer(inputs)
    labels = paddle.randn([10, 1], 'float32')
    loss = loss_fn(outputs, labels)

    if print_result is True:
        print("loss:", loss.numpy())

    loss.backward()

    adam.step()
    adam.clear_grad()

# Usage 1: only pass function.
# If your training method no need any argument, and
# use all visible devices for parallel training.
if __name__ == '__main__':
    dist.spawn(train)

# Usage 2: pass function and arguments.
# If your training method need some arguments, and
# use all visible devices for parallel training.
if __name__ == '__main__':
    dist.spawn(train, args=(True,))

# Usage 3: pass function, arguments and nprocs.
# If your training method need some arguments, and
# only use part of visible devices for parallel training.
# If your machine hold 8 cards {0,1,2,3,4,5,6,7},
# this case will use cards {0,1}
if __name__ == '__main__':
    dist.spawn(train, args=(True,), nprocs=2)

# Usage 4: pass function, arguments, nprocs and gpus.
# If your training method need some arguments, and
# only use part of visible devices for parallel training, 
# you can pass `xpus` to select the XPU cards you want to use. 
# For example, this case will use cards {4,5}  if your machine hold more than 6 cards.
if __name__ == '__main__':
    dist.spawn(train, args=(True,), nprocs=2, xpus='4,5')
```
The only  difference in API interface between xpus and gpus is usage 4.

model benchmarks:
- BERT
  结果测量方法：取第一个epoch第10-110step的平均
  数据单位：step/s
  卡类型：K200
  单卡batch size=16

卡数/精度 | 1N1C | 1N2C | 1N4C | 1N8C | 1N16C
-- | -- | -- | -- | -- | --
FP32 | 1.46x(1.0x) | 1.42*2=2.88(1.9452x) | 0.86*4=3.44(2.3561x) | 0.77*8= 6.16(4.3380x) | 0.52*16= 8.32(5.6986x)

- ResNet
动态图多进程多卡
结果测量方法：取第一个epoch第10-110step的平均。
数据单位：ips
卡类型：K200
单卡batch size=32 

卡数/精度 | 1N1C | 1N2C | 1N4C | 1N8C | 1N16C
-- | -- | -- | -- | -- | --
FP32 | 52.56127 | 46.21590*2=92.4318(1.7585x) | 41.66224*4=166.64896(3.1705x) | 36.2901*8=290.32007=(5.5234x) | shm error

### CODE REVIEW

File: paddle/fluid/imperative/reducer.cc [hunk 1]
```diff
@@ -640,56 +641,80 @@ void Reducer::MarkGroupReady(size_t group_index) {
     return;
   }
 
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    multi_device_op_count_ = 0;
```
Comment (wangxicoding) [inline paddle/fluid/imperative/reducer.cc]: 这个放在Reducer构造函数中吧

File: paddle/fluid/imperative/reducer.cc [hunk 2]
```diff
@@ -640,56 +641,80 @@ void Reducer::MarkGroupReady(size_t group_index) {
...
-      // Select common commstream to split tensors
-      // group.dense_contents_ ---> group.dense_tensors
-      group.SplitTensors(*parallel_ctx_->GetDeviceContext(run_order));
-    }
+      {
+        std::lock_guard<std::mutex> lock(mutex_);
+        multi_device_op_count_ -= 1;  // lock
+        cv_.notify_all();
+      }
+      // cnt -= 1; // lock
```
Comment (wangxicoding) [inline paddle/fluid/imperative/reducer.cc]: delete

File: python/paddle/distributed/spawn.py [hunk 3]
```diff
@@ -407,8 +452,10 @@ def train(print_result=False):
         if device == 'cpu':
             # TODO: not supports cpu parallel now
             nprocs = _cpu_num()
-        else:
+        elif device == 'gpu':
             nprocs = core.get_cuda_device_count()
+        else:
```
Comment (wangxicoding) [inline python/paddle/distributed/spawn.py]: elif device == 'xpu'

**Review by wangxicoding — CHANGES_REQUESTED** [review]

File: paddle/fluid/imperative/reducer.h [hunk 4]
```diff
@@ -187,6 +187,13 @@ class Reducer {
   bool has_marked_unused_vars_{false};
   bool find_unused_vars_{false};
   bool all_group_ready_{false};
+
+  // std::vector<std::unique_ptr<::ThreadPool>> pool_;
+  // ::ThreadPool comm_pool_;
+  ::ThreadPool multi_device_op_pool_;
```
Comment (wangxicoding) [inline paddle/fluid/imperative/reducer.h]: 名字就改成上面的comm_pool，或者想一个更好的。再加些注释，这个线程用于调度通信allreduce

File: paddle/fluid/imperative/reducer.h [hunk 5]
```diff
@@ -187,6 +187,13 @@ class Reducer {
...
+  uint32_t multi_device_op_count_;
```
Comment (wangxicoding) [inline paddle/fluid/imperative/reducer.h]: 这个也改一下，comm_op_count，或者想个更好的

File: paddle/fluid/imperative/reducer.cc [hunk 6]
```diff
@@ -640,56 +641,80 @@ void Reducer::MarkGroupReady(size_t group_index) {
...
+    multi_device_op_count_ = 0;
+  }
+
   for (; next_group_ < groups_.size() && groups_[next_group_].pending_ == 0;
        ++next_group_) {
     auto &group = groups_[next_group_];
     int run_order = next_group_ % nrings_;
 
-    // For CUDA or XPU, compute_stream --> comm_stream.
+    // For CUDA or XPU, compute_stream --event--> comm_stream.
```
Comment (QingshuChen) [inline paddle/fluid/imperative/reducer.cc]: event也加进去了吗？

Comment (wangxicoding) [reply paddle/fluid/imperative/reducer.cc]: 没有，通信库现在还是阻塞的，加进去也没有效果

File: paddle/fluid/imperative/reducer.h (hunk 5 above)
Comment (vslyu) [reply paddle/fluid/imperative/reducer.h]: just  comm_op_count_?

File: python/paddle/distributed/spawn.py (hunk 3 above)
Comment (vslyu) [reply python/paddle/distributed/spawn.py]: done.

File: paddle/fluid/imperative/reducer.cc (hunk 2 above)
Comment (vslyu) [reply paddle/fluid/imperative/reducer.cc]: done.

File: paddle/fluid/imperative/reducer.cc (hunk 1 above)
Comment (vslyu) [reply paddle/fluid/imperative/reducer.cc]: done.

File: paddle/fluid/imperative/reducer.h (hunk 4 above)
Comment (vslyu) [reply paddle/fluid/imperative/reducer.h]: just comm_pool_?

File: paddle/fluid/imperative/reducer.cc [hunk 7]
```diff
@@ -645,13 +656,73 @@ void Reducer::MarkGroupReady(size_t group_index) {
...
+        }
+      }
+
+      {
+        std::lock_guard<std::mutex> lock(mutex_);
+        comm_op_count_ -= 1;  // lock
+        cv_.notify_all();
+      }
+    });
+#else
```
Comment (wangxicoding) [inline paddle/fluid/imperative/reducer.cc:680]: GPU的WaitCompute丟了

**Review by wangxicoding — CHANGES_REQUESTED** [review]

File: paddle/fluid/imperative/reducer.cc (hunk 7 above)
Comment (vslyu) [reply paddle/fluid/imperative/reducer.cc:680]: done, add it again.

File: python/paddle/distributed/spawn.py [hunk 8]
```diff
@@ -117,39 +117,75 @@ def _get_subprocess_env_list(nprocs, options):
...
+                    "correct `nprocs` and `gpus` arguments are passed." %
+                    (len(selected_gpu_list), nprocs))
+            for card_id in selected_gpu_list:
+                if card_id not in env_devices_list:
+                    raise ValueError("The selected gpu card %s cannot found in "
+                                     "CUDA_VISIBLE_DEVICES (%s)." %
+                                     (card_id, ",".join(env_devices_list)))
+
+    if core.is_compiled_with_xpu():
+        args.selected_gpus = options.get('xpus', None)
```
Comment (chenwhql) [inline python/paddle/distributed/spawn.py]: xpu也使用的是`args.selected_gpus`吗？感觉代码读起来有点乱

File: python/paddle/distributed/spawn.py [hunk 9]
```diff
@@ -117,39 +117,75 @@ def _get_subprocess_env_list(nprocs, options):
...
+                    "`nprocs` argument is passed or the environment variable "
+                    "`XPU_VISIBLE_DEVICES` is correctly configured." %
+                    (len(env_devices_list), nprocs))
+            args.selected_gpus = ",".join(
+                [str(env_devices_list[x]) for x in range(0, nprocs)])
+        else:
+            selected_gpu_list = args.selected_gpus.split(',')
+            if len(selected_gpu_list) != nprocs:
+                raise ValueError(
+                    "The number of selected gpus(%s) is not equal to "
```
Comment (chenwhql) [inline python/paddle/distributed/spawn.py]: `gpus` -> `xpus`?

File: python/paddle/distributed/spawn.py [hunk 10]
```diff
@@ -117,39 +117,75 @@ def _get_subprocess_env_list(nprocs, options):
...
+                    "the number of spawn processes(%d), please ensure that the "
+                    "correct `nprocs` and `gpus` arguments are passed." %
```
Comment (chenwhql) [inline python/paddle/distributed/spawn.py]: `gpus` -> `xpus`?

File: python/paddle/distributed/spawn.py [hunk 11]
```diff
@@ -117,39 +117,75 @@ def _get_subprocess_env_list(nprocs, options):
...
+                    (len(selected_gpu_list), nprocs))
+            for card_id in selected_gpu_list:
+                if card_id not in env_devices_list:
+                    raise ValueError("The selected gpu card %s cannot found in "
```
Comment (chenwhql) [inline python/paddle/distributed/spawn.py]: `gpu` -> `xpu`?

File: paddle/fluid/imperative/reducer.cc [hunk 12]
```diff
@@ -645,13 +656,74 @@ void Reducer::MarkGroupReady(size_t group_index) {
...
+// so we expose WaitCompute() interface and call
+// it here.
+#ifdef PADDLE_WITH_XPU_BKCL
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      comm_op_count_ += 1;  // lock
+    }
+
+    comm_pool_.enqueue([&] {
+      parallel_ctx_->WaitCompute(run_order);
```
Comment (wangxicoding) [inline paddle/fluid/imperative/reducer.cc]: SetXPUDevice呢

**Review by wangxicoding — CHANGES_REQUESTED** [review]

File: python/paddle/distributed/spawn.py (hunk 10 above)
Comment (vslyu) [reply python/paddle/distributed/spawn.py]: done.

File: python/paddle/distributed/spawn.py (hunk 8 above)
Comment (vslyu) [reply python/paddle/distributed/spawn.py]: Done, uniformally define args.selected_devices for multi gpus training or multi xpus training, delete args.selected_gpus.

File: python/paddle/distributed/spawn.py (hunk 11 above)
Comment (vslyu) [reply python/paddle/distributed/spawn.py]: done.

File: paddle/fluid/imperative/reducer.cc [hunk 13]
```diff
@@ -292,12 +292,23 @@ Reducer::Reducer(const std::vector<std::shared_ptr<imperative::VarBase>> &vars,
...
                  bool find_unused_vars)
+#ifdef PADDLE_WITH_XPU_BKCL
+    : vars_(vars),
+      group_indices_(group_indices),
+      is_sparse_gradient_(is_sparse_gradient),
+      parallel_ctx_(parallel_ctx),
+      group_size_limits_(group_size_limits),
+      find_unused_vars_(find_unused_vars),
+      comm_pool_(1),
+      comm_op_count_(0) {
```
Comment (hutuxian) [inline paddle/fluid/imperative/reducer.cc]: only wrap the added code

File: paddle/fluid/imperative/reducer.cc (hunk 12 above)
Comment (vslyu) [reply paddle/fluid/imperative/reducer.cc]: done, add it.

File: paddle/fluid/imperative/reducer.cc (hunk 13 above)
Comment (vslyu) [reply paddle/fluid/imperative/reducer.cc]: done.

File: paddle/fluid/imperative/reducer.cc [hunk 14]
```diff
@@ -645,51 +652,84 @@ void Reducer::MarkGroupReady(size_t group_index) {
     auto &group = groups_[next_group_];
     int run_order = next_group_ % nrings_;
 
+    auto place = parallel_ctx_->GetDeviceContext(run_order)->GetPlace();
+
```
Comment (ForFishes) [inline paddle/fluid/imperative/reducer.cc]: 这里为什么需要拿place，place不是已经有了吗？这里涉及gpu和xpu的混合通信？

File: paddle/fluid/imperative/reducer.cc [hunk 15]
```diff
@@ -645,51 +652,84 @@ void Reducer::MarkGroupReady(size_t group_index) {
...
+      FusedAllReduceSchedule(run_order, group);
+#else
+      PADDLE_THROW(
+          platform::errors::PreconditionNotMet("Not compiled with NCCL."));
+#endif
+    }
+  }
+}
+
+void Reducer::FusedAllReduceSchedule(int run_order, Group group) {
```
Comment (ForFishes) [inline paddle/fluid/imperative/reducer.cc]: 换成常量引用，不然影响性能。

File: paddle/fluid/imperative/reducer.cc (hunk 14 above)
Comment (wangxicoding) [reply paddle/fluid/imperative/reducer.cc]: reducer.h 里已经有 _place了，不用获取了。

File: paddle/fluid/imperative/reducer.cc [hunk 16]
```diff
@@ -645,51 +652,84 @@ void Reducer::MarkGroupReady(size_t group_index) {
...
     int run_order = next_group_ % nrings_;
 
+    auto place = parallel_ctx_->GetDeviceContext(run_order)->GetPlace();
+
     // For CUDA or XPU, compute_stream --> comm_stream.
     // For CPU, do nothing.
     // NOTE. Because concat uses the comm_stream,
     // so we expose WaitCompute() interface and call
     // it here.
     parallel_ctx_->WaitCompute(run_order);
```
Comment (wangxicoding) [inline paddle/fluid/imperative/reducer.cc:659]: 这个后续试一下WaitCompute放到这里，还是放到allreduce里面的性能好

File: paddle/fluid/imperative/reducer.cc [hunk 17]
```diff
@@ -645,51 +652,84 @@ void Reducer::MarkGroupReady(size_t group_index) {
...
-      if (group.sparse_contents_ != nullptr) {
-        VLOG(3) << "sparse group [" << next_group_
-                << "] start allreduce in ring[" << run_order << "]";
-        group.DivNRanks(*parallel_ctx_->GetDeviceContext(run_order), nranks_);
-        parallel_ctx_->AllReduceByStream(
-            *group.sparse_contents_, group.sparse_contents_, run_order, false);
-      } else {
-        VLOG(3) << "The sparse group[" << next_group_
-                << "] has no var to allreduce";
+    if (paddle::platform::is_xpu_place(place)) {
```
Comment (wangxicoding) [inline paddle/fluid/imperative/reducer.cc]: 可以不用判断place了，直接BKCL宏定义线程调度。其余正常调度

File: paddle/fluid/imperative/reducer.cc [hunk 18]
```diff
@@ -645,51 +652,84 @@ void Reducer::MarkGroupReady(size_t group_index) {
...
+#ifdef PADDLE_WITH_XPU_BKCL
+      {
+        std::lock_guard<std::mutex> lock(mutex_);
+        comm_op_count_ += 1;  // lock
       }
+
+      comm_pool_->enqueue([&] {
+        auto dev_id = BOOST_GET_CONST(platform::XPUPlace, place).device;
+        platform::SetXPUDeviceId(dev_id);
+        FusedAllReduceSchedule(run_order, group);
```
Comment (wangxicoding) [inline paddle/fluid/imperative/reducer.cc]: 线程调度，记个TOTO，后续加上try cache。否则出异常了主线程不知道还一直在跑

**Review by Xreki — DISMISSED** [review]
LGTM for the modification of compiler.py

**Review by wangxicoding — DISMISSED** [review]
LGTM

**Review by ForFishes — APPROVED** [review]
LGTM

**Review by wangxicoding — APPROVED** [review]

**Review by Xreki — APPROVED** [review]

**Review by luotao1 — APPROVED** [review]

**Pull request merged**